java/ddlogapi.c

```diff
@@ -13,6 +13,74 @@
 #include "ddlog.h"
 #include "ddlog_log.h"
 
+/* Error message returned by DDlog. */
+_Thread_local char* err_msg = NULL;
+
+/* 
+ * `vasprintf` implementation copied from
+ * https://github.com/littlstar/asprintf.c (MIT license)
+ */
+static int _vasprintf (char **str, const char *fmt, va_list args) {
+  int size = 0;
+  va_list tmpa;
+
+  va_copy(tmpa, args);
+  size = vsnprintf(NULL, 0, fmt, tmpa);
+  va_end(tmpa);
+
+  if (size < 0) { return -1; }
+
+  // alloc with size plus 1 for `\0'
+  *str = (char *) malloc(size + 1);
+
+  if (NULL == *str) { return -1; }
+
+  size = vsprintf(*str, fmt, args);
+  return size;
+}
+
+/* 
+ * Callback invoked by DDlog on error.
+ *
+ * Record error message in a thread-local variable so that it can later be
+ * wrapped in an exception.
+ */
+static void eprintln(const char* msg) {
```

Comment:
I hope this is race-free...

Reply:
Well, it uses thread-local state, so hopefully should be.  The dodginess here is that we rely on DDlog to print an error on failure, which is an invariant we should be careful to observe.